### PR TITLE
fix: add customFormatErrorFn, unhandled GraphQLErrors should fail e2e tests

### PR DESCRIPTION
**Pull request by tgriesser**
We should never have unhandled GraphQL errors - make sure these get propagated during e2e tests

### CODE REVIEW

**Review by lmiller1990 — APPROVED** [review]
Added a test

https://github.com/cypress-io/cypress/pull/19937/files#diff-cd8a26c958116ef64041aa84b46f14e7ba2f779e01e5994fc7b93f2a4d84aa5d

File: packages/frontend-shared/src/graphql/urqlClient.ts [hunk 1]
```diff
@@ -99,6 +99,12 @@ export function makeUrqlClient (config: UrqlClientConfig): Client {
           })
         }
 
+        if (['test', 'production'].includes(process.env.NODE_ENV as string)) {
+          // In production, we want to complain loudly if there is
+          // ever a unhandled GraphQL error, since that's a bug in Cypress.
+          throw error
+        }
```
Comment (tgriesser) [inline packages/frontend-shared/src/graphql/urqlClient.ts:106]: @lmiller1990 why was this added? I haven't tested but seem to remember that re-throwing the error will cancel the observable pipeline, which we definitely don't want in production.

**Review by tgriesser — COMMENTED** [review]
I'm not sure that we want to fatally error if there's a field error. We might want to log it, but probably not re-throw it.

File: packages/frontend-shared/src/graphql/urqlClient.ts (hunk 1 above)
Comment (lmiller1990) [reply packages/frontend-shared/src/graphql/urqlClient.ts:106]: I could be missing something, but isn't the whole point of this for unhandled GraphQL errors to fail the e2e test? Without throwing this, those errors will not cause a test failure.

If there's an error on a field, wouldn't this be considered a bug in Cypress? We should not have unhandled errors.


I could be misunderstanding the direction this PR was heading.

Comment (tgriesser) [reply packages/frontend-shared/src/graphql/urqlClient.ts:106]: Yeah, it’s capturing them on the backend and we have a global hook we’re calling in the after to assert there were none, so the test will pass but the suite will fail.

It’s not reliable to handle them in the front end because the test may have passed before the error is seen in the client.

Comment (lmiller1990) [reply packages/frontend-shared/src/graphql/urqlClient.ts:106]: If that was the intention, I was not able to see the suite actually failing with throwing the error. Was your code working as intended? Eg, I didn't see a test or any way to actually verify that the suites would fail, so I added an endpoint that always errors, and visited it in e2e. Despite the error, the suite still passed (before adding the code that I did).

If we don't throw an error, uncaught GraphQL exceptions will *not* cause the suite to fail, at least when I tried it locally. That's why I added the code I did.

I think there's some confusion, let's sync on this.

Comment (tgriesser) [reply packages/frontend-shared/src/graphql/urqlClient.ts:106]: I thought it was? In [this commit](https://github.com/cypress-io/cypress/pull/19937/commits/068d37d2c7e4b5dc33fea30f12c3d26e7243b326) I was seeing it catch errors:

https://app.circleci.com/pipelines/github/cypress-io/cypress/30530/workflows/52fd9a35-209d-4e86-a931-ef499a8f4c9a/jobs/1186226

https://app.circleci.com/pipelines/github/cypress-io/cypress/30530/workflows/52fd9a35-209d-4e86-a931-ef499a8f4c9a/jobs/1186222

Comment (lmiller1990) [reply packages/frontend-shared/src/graphql/urqlClient.ts:106]: I see - I wonder if we can test **that** it errors and causes a fail? That's what I was going for.

Maybe we just revert my commit and merge, then?

**Comment by MuazOthman** [comment]
@tgriesser do we need any of this code?

**Review by jennifer-shehane — CHANGES_REQUESTED** [review]
@tgriesser Update base branch to be against develop before merging.
